torch/fx/passes/split_module.py

```diff
@@ -194,6 +196,76 @@ def instantiate_node_partition_mapping(node):
         partition.node_names.append(node.name)
         node._fx_partition = partition_name
 
+    # Global State Nodes are nodes which by their global state effects,
+    # "taint" all downstream nodes while they are active.
+    GLOBAL_STATE_NODES = [
+        torch.amp._enter_autocast,
+        torch.amp._exit_autocast,
+        torch._C._set_grad_enabled
+    ]
+
+    # For grad regions:
+    # ------------------------
+    # 1. first region: we do nothing
+    # 2. subsequent regions: we insert the set_grad at the beginning
+    grad_regions: OrderedDict[Node, Set[int]] = OrderedDict()
+
+    # For autocast regions:
+    # ------------------------
+    # 1. first region: we will only insert the _exit at the end
+    # 2. intermediate regions: we will insert both the
+    #    _enter at the beginning and _exit at the end
+    # 3. last region: we will only insert _enter at the beginning
+    # We will do so in the order in which the autocasts were instantiated.
+    autocast_regions: OrderedDict[Node, Set[int]] = OrderedDict()
+    autocast_exits: Dict[Node, Optional[Node]] = {}
+
+    active_grad = None
+    active_autocasts = set()
+
+    for node in m.graph.nodes:
+        if node.op in ["placeholder", "get_attr", "output"]:
+            continue
+
+        instantiate_node_partition_mapping(node)
+
+        if node.op == "call_function" and node.target in GLOBAL_STATE_NODES:
+            if node.target == torch._C._set_grad_enabled:
+                assert len(node.args) == 1
+                assert isinstance(node.args[0], bool)
+                active_grad = node
+                grad_regions[active_grad] = set({split_callback(node)})
+            elif node.target == torch.amp._enter_autocast:
+                # Should all be python constants
+                assert all(not isinstance(arg, Node) for arg in node.args)
+                active_autocasts.add(node)
+                autocast_regions[node] = set({split_callback(node)})
+                autocast_exits[node] = None
+            elif node.target == torch.amp._exit_autocast:
+                assert len(node.args) == 1
+                autocast_regions[node.args[0]].add(split_callback(node))
+                active_autocasts.remove(node.args[0])
+                autocast_exits[node.args[0]] = node
+
+        if active_grad is not None:
+            grad_regions[active_grad].add(split_callback(node))
+
+        for a in active_autocasts:
```

Comment:
would you add a test case that covers more than one / nested autocast regions?  (and possibly also grad region in the same graph as autocast region)

Reply:
Yes, I thought of doing so and to properly test we need a custom model which returns intermediate outputs whose requires_grad / dtype should differ based on which layer it is in (corresponding to which ctx manager is active in a given layer). I'll add it.

Reply:
Added a nested test with example output:

```python
---orig graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=2] = placeholder[target=L_inputs_]
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_0 : [num_users=1] = call_module[target=L__self___net1_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net1_1 : [num_users=1] = call_module[target=L__self___net1_1](args = (%l__self___net1_0,), kwargs = {})
    %l__self___net1_2 : [num_users=1] = call_module[target=L__self___net1_2](args = (%l__self___net1_1,), kwargs = {})
    %l__self___net1_3 : [num_users=1] = call_module[target=L__self___net1_3](args = (%l__self___net1_2,), kwargs = {})
    %l__self___net1_4 : [num_users=1] = call_module[target=L__self___net1_4](args = (%l__self___net1_3,), kwargs = {})
    %intermediates_1 : [num_users=2] = call_module[target=L__self___net1_5](args = (%l__self___net1_4,), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_0 : [num_users=1] = call_module[target=L__self___net2_0](args = (%intermediates_1,), kwargs = {})
    %l__self___net2_1 : [num_users=1] = call_module[target=L__self___net2_1](args = (%l__self___net2_0,), kwargs = {})
    %l__self___net2_2 : [num_users=1] = call_module[target=L__self___net2_2](args = (%l__self___net2_1,), kwargs = {})
    %intermediates_2 : [num_users=1] = call_module[target=L__self___net2_3](args = (%l__self___net2_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %l__self___net3_0 : [num_users=1] = call_module[target=L__self___net3_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net3_1 : [num_users=1] = call_module[target=L__self___net3_1](args = (%l__self___net3_0,), kwargs = {})
    %l__self___net3_2 : [num_users=1] = call_module[target=L__self___net3_2](args = (%l__self___net3_1,), kwargs = {})
    %l__self___net3_3 : [num_users=1] = call_module[target=L__self___net3_3](args = (%l__self___net3_2,), kwargs = {})
    %l__self___net3_4 : [num_users=1] = call_module[target=L__self___net3_4](args = (%l__self___net3_3,), kwargs = {})
    %outputs : [num_users=1] = call_module[target=L__self___net3_5](args = (%l__self___net3_4,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (intermediates_1, intermediates_2, outputs)

---split graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=2] = placeholder[target=L_inputs_]
    %submod_0 : [num_users=1] = call_module[target=submod_0](args = (%l_inputs_,), kwargs = {})
    %submod_1 : [num_users=2] = call_module[target=submod_1](args = (%submod_0,), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%submod_1, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_1, 1), kwargs = {})
    %submod_2 : [num_users=2] = call_module[target=submod_2](args = (%getitem, %l_inputs_), kwargs = {})
    %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_2, 0), kwargs = {})
    %getitem_3 : [num_users=1] = call_function[target=operator.getitem](args = (%submod_2, 1), kwargs = {})
    %submod_3 : [num_users=1] = call_module[target=submod_3](args = (%getitem_2,), kwargs = {})
    return (getitem_1, getitem_3, submod_3)

---submod_0 graph---
graph():
    %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_0 : [num_users=1] = call_module[target=L__self___net1_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net1_1 : [num_users=1] = call_module[target=L__self___net1_1](args = (%l__self___net1_0,), kwargs = {})
    %l__self___net1_2 : [num_users=1] = call_module[target=L__self___net1_2](args = (%l__self___net1_1,), kwargs = {})
    %l__self___net1_3 : [num_users=1] = call_module[target=L__self___net1_3](args = (%l__self___net1_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return l__self___net1_3

---submod_1 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net1_3 : [num_users=1] = placeholder[target=l__self___net1_3]
    %l__self___net1_4 : [num_users=1] = call_module[target=L__self___net1_4](args = (%l__self___net1_3,), kwargs = {})
    %l__self___net1_5 : [num_users=2] = call_module[target=L__self___net1_5](args = (%l__self___net1_4,), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_0 : [num_users=1] = call_module[target=L__self___net2_0](args = (%l__self___net1_5,), kwargs = {})
    %l__self___net2_1 : [num_users=1] = call_module[target=L__self___net2_1](args = (%l__self___net2_0,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (l__self___net2_1, l__self___net1_5)

---submod_2 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %_enter_autocast_1 : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net2_1 : [num_users=1] = placeholder[target=l__self___net2_1]
    %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
    %l__self___net2_2 : [num_users=1] = call_module[target=L__self___net2_2](args = (%l__self___net2_1,), kwargs = {})
    %l__self___net2_3 : [num_users=1] = call_module[target=L__self___net2_3](args = (%l__self___net2_2,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
    %l__self___net3_0 : [num_users=1] = call_module[target=L__self___net3_0](args = (%l_inputs_,), kwargs = {})
    %l__self___net3_1 : [num_users=1] = call_module[target=L__self___net3_1](args = (%l__self___net3_0,), kwargs = {})
    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return (l__self___net3_1, l__self___net2_3)

---submod_3 graph---
graph():
    %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
    %l__self___net3_1 : [num_users=1] = placeholder[target=l__self___net3_1]
    %l__self___net3_2 : [num_users=1] = call_module[target=L__self___net3_2](args = (%l__self___net3_1,), kwargs = {})
    %l__self___net3_3 : [num_users=1] = call_module[target=L__self___net3_3](args = (%l__self___net3_2,), kwargs = {})
    %l__self___net3_4 : [num_users=1] = call_module[target=L__self___net3_4](args = (%l__self___net3_3,), kwargs = {})
    %l__self___net3_5 : [num_users=1] = call_module[target=L__self___net3_5](args = (%l__self___net3_4,), kwargs = {})
    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
    return l__self___net3_5

```

Reply:
>                 with self.ctx_manager_1():
                    intermediates_1 = self.net1(inputs)
                    with self.ctx_manager_2():
                        intermediates_2 = self.net2(intermediates_1)
                    outputs = self.net3(inputs)


~~Assuming the above graph is from this model, isn't it wrong? I would expect the exits to be in reverse order to the enters~~

ok nvm they are correct, but, confusingly the `_exit_autocast` node exits the `_enter_autocast_1`, while exit1 exits enter0...  Not sure if that's worth fixing (would it be hard?)

`    %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast_1,), kwargs = {})
`

`    %_exit_autocast_1 : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
`

Reply:
Yes, this is just how the fx graph names things when there are name collisions. I don't think it's worth fixing though it is confusing.

Reply:
I'm going to resolve this - it's just an artifact of fx name collisions and is IMO not a useability concern as long as one scruitinizes the graph properly.